xcat-inventory/xcclient/allien/regexutils.py

```diff
@@ -0,0 +1,95 @@
+###############################################################################
+# IBM(c) 2019 EPL license http://www.eclipse.org/legal/epl-v10.html
+###############################################################################
+
+# -*- coding: utf-8 -*-
+
+import re
+
+def extract_bracketed(input_str):
+    """extract bracketed expression into dict
+       input_str example: 10.($1+1).0.($2+3)
+       output example:
+        { 'prev': '10.',
+          'curr': '($1+1)',
+          'next': '.0.($2+3)'
+        }
+
+    """
+    result = {}
+    pre=''
+    cur=''
+    nxt=''
+    sign = 0
+    loop=0
+    for ch in input_str:
+        if sign is not 2:
+            if ch == '(':
+                sign=1
+            elif (ch == ')'):
+                sign=2
+                loop+=1
+                if loop is 1:
+                    cur += ch
+                    ch=''
+        if sign is 1:
+            cur += ch
+        elif sign is 2 and ch:
+            nxt += ch
+        elif sign is 0:
+            pre += ch
+    result['prev']=pre
+    result['curr']=cur
+    result['next']=nxt
+    return result
+
+def trans_regex_attrs(node,attr):
```

Comment:
Better to define a method to check (attr), and return a tuple:  ( 0/1/2, None or RE object )